File: yggy/observable/observable.py
import abc
import uuid
from typing import TYPE_CHECKING, Any, Callable
from weakref import WeakKeyDictionary

from ..comm import ReceiverFn_t, create_message
from ..logging import get_logger
from ..utils.functools import noop
from .messages import OBSERVABLE_CHANGE_MSG, ChangeMessage
from .schema import ObservableSchema

if TYPE_CHECKING:
    from .observable_network import ObservableNetwork

__all__ = [
    "Observable",
    "Primitive",
    "get",
]

logger = get_logger(f"{__name__}")

type Primitive = None | bool | int | float | str


def get[T: Primitive](obs: "Observable[T] | T") -> T:
    if isinstance(obs, Observable):
        return obs.get()
    return obs


class Observable[T: Primitive](abc.ABC):
    __network: "ObservableNetwork | None"
    __id: str
    __value: T
    __coerce: Callable[[Any], T]
    __validate: Callable[[T], T]
    __receivers: WeakKeyDictionary[ReceiverFn_t[Any], ReceiverFn_t[Any]]
    __last_change: str | None

    def __init__(
        self,
        __value: T,
        *,
        coerce: Callable[[Any], T] | None = None,
        validate: Callable[[T], T] | None = None,
    ) -> None:
        super().__init__()

        self.__network = None
        self.__id = uuid.uuid4().hex
        self.__value = __value
        self.__coerce = coerce or type(__value) if __value is not None else noop
        self.__validate = validate or noop
        self.__receivers = WeakKeyDictionary()
        self.__last_change = None

    def __json__(self) -> ObservableSchema[T]:
        return {"data_id": self.id, "value": self.get()}

    @property
    def id(self) -> str:
        return self.__id

    def watch(
        self, __fn: ReceiverFn_t[ChangeMessage[T]]
    ) -> ReceiverFn_t[ChangeMessage[T]]:
        if __fn in self.__receivers:
            return __fn

        def __recv_watch(__change: ChangeMessage[T]) -> None:
            if __change["data_id"] != self.id:
                return
            __fn(__change)

        self.__receivers[__fn] = __recv_watch

        if self.__network is not None:
            self.__network.comm.recv(OBSERVABLE_CHANGE_MSG, __recv_watch)

        return __fn

    def unwatch(
        self, __fn: ReceiverFn_t[ChangeMessage[T]]
    ) -> ReceiverFn_t[ChangeMessage[T]]:
        if __fn in self.__receivers:
            if self.__network is not None:
                self.__network.comm.unrecv(
                    OBSERVABLE_CHANGE_MSG, self.__receivers[__fn]
                )
            del self.__receivers[__fn]
        return __fn

    def _register(self, __network: "ObservableNetwork") -> None:
        self.__network = __network
        for receiver in self.__receivers.values():
            self.__network.comm.recv(OBSERVABLE_CHANGE_MSG, receiver)

    def get(self) -> T:
        return self._get()

    def set(self, __value: T) -> None:
        try:
            __value = self.__validate(self.__coerce(__value))
            old_value = self.get()
        except BaseException as e:
            logger.error(e)
            old_value = __value = self.get()

        self._set(__value)
        self._notify_change(old_value, __value)

    def _get(self) -> T:
        return self.__value

    def _set(self, __value: T) -> None:
        self.__value = __value

    def _notify_change(self, old_value: T, new_value: T) -> None:
        if self.__network is None:
            return
        change = create_message(
            ChangeMessage[T],
            {
                "data_id": self.id,
                "new_value": new_value,
                "old_value": old_value,
            },
        )
        # TODO: look into message revoking more, as it might still work
        # TODO: followup: LazyMessage which evaluates the message when popped from queue
        # self.__network.send_change(self, change, revoke=self.__last_change)
        self.__network.send_change(self, change)
        self.__last_change = change.get("message_id", None)

File: yggy/observable/observable.ts
import { ReceiverFn_t, create_message } from "../comm/__init__.js";
import { weakref } from "../utils/__init__.js";
import { ChangeMessage, OBSERVABLE_CHANGE_MSG } from "./messages.js";
import { ObservableNetwork } from "./observable_network.js";
import { ObservableSchema } from "./schema.js";

type ObservableKwds = {
    local?: boolean;
}

export function get<T>(obs: Observable<T> | T): T {
    if (obs instanceof Observable) {
        return obs.get();
    }
    return obs;
}

export class Observable<T> {
    #network?: ObservableNetwork;
    #id: string;
    #value: T;
    #receivers: weakref.IterableWeakMap<ReceiverFn_t, ReceiverFn_t>;
    #local: boolean;

    constructor(__id: string, __value: T, __kwds: ObservableKwds = {}) {
        this.#id = __id;
        this.#value = __value;
        this.#receivers = new weakref.IterableWeakMap();
        this.#local = __kwds.local ?? false;
    }

    get id(): string {
        return this.#id;
    }

    get network(): ObservableNetwork | void {
        return this.#network;
    }

    static from_schema<T>(__schema: ObservableSchema<T>): Observable<T> {
        return new Observable(__schema.data_id, __schema.value);
    }

    register(__network: ObservableNetwork): void {
        this.#network = __network;
        for (let fn of this.#receivers.values()) {
            this.#network.comm.recv(OBSERVABLE_CHANGE_MSG, fn);
        }
    }

    get(): T {
        return this.#value;
    }

    set(__value: T): void {
        const old_value = this.#value;
        this.#value = __value;

        this.#notify_change(old_value, __value);
    }


    watch(__fn: ReceiverFn_t): void {
        const __recv_change = (__change: ChangeMessage<any>) => {
            if (__change["data_id"] != this.id) { return; }
            __fn(__change);
        };

        this.#receivers.set(__fn, __recv_change);
        this.#network?.comm.recv(OBSERVABLE_CHANGE_MSG, __recv_change);
    }

    #notify_change<T>(old_value: T, new_value: T): void {
        const change = create_message<ChangeMessage<T>>({
            data_id: this.id,
            new_value: new_value,
            old_value: old_value,
        });

        if (this.#local) {
            this.#network?.notify_change(change);
        }
        else {
            this.#network?.send_change(change);
        }
    }

    __recv_change(change: ChangeMessage<T>): void {
        this.set(change.new_value);
    }
}
